akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -401,16 +401,40 @@ private[stream] object Collect {
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
     new GraphStageLogic(shape) with InHandler with OutHandler {
+      private var buffer: OptionVal[T] = OptionVal.none
+      private val contextPropagation = ContextPropagation()
+
       private var left: Long = count
 
+      override def preStart(): Unit = pull(in)
+
       override def onPush(): Unit = {
         if (left > 0) {
           left -= 1
           pull(in)
-        } else push(out, grab(in))
+        } else {
+          val elem = grab(in)
+          if (isAvailable(out)) {
+            push(out, elem)
+          } else {
+            buffer = OptionVal.Some(elem)
+            contextPropagation.suspendContext()
+          }
+        }
```

Comment:
So the buffer will only be used once, first push after dropping `count`, then it switches over to the old unbuffered/detached behavior. Could maybe be cleaner with a separate handler eagerly pulling/dropping and then switch once first element reached (not entirely sure)?